protocol/smart-contract-versioning.md

```diff
@@ -0,0 +1,86 @@
+# Purpose
+
+This design doc proposes a new smart contracts versioning scheme to improve clarity and more closely align with typical semantic versioning practices.
+
+# Summary
+
+The proposed versioning scheme for contracts will follow a true semver format, with major, minor, and patch versions, as well as beta, release candidate, and feature tags.
+The monorepo releases will also follow semver, with the version bump determined by the highest version bump of any individual contract modified in the release.
+
+# Problem Statement + Context
+
+There are two facets to contract versioning: the versioning of monorepo releases, and the versioning of individual contracts.
+
+- **Monorepo Releases**:  The current versioning scheme of monorepo contract releases does not follow a true semver. For example, the Fault Proofs release introduced breaking changes by changing the ABI, and we are releasing this as a minor version bump from `op-contracts/v1.3.0` to `op-contracts/v1.4.0`. This was not an intentional decision. We likely slid into this path because node releases cannot bump the major version (for go-specific reasons), and our current [contracts release process](https://www.notion.so/oplabs/Minimal-Contracts-Release-Process-ef8e91b1b13944e7a7f632a824895764?pvs=4) does not specify how to determine the version bump for an upcoming monorepo release.
```

Comment:
I notice that everything is `op-contracts`, what if we want to release L1 and L2 contracts differently? Or generally different subpackages of contracts. I would like to enshrine the idea of tying L2 predeploys to hardforks (related to [this](https://github.com/ethereum-optimism/specs/pull/194), proposal: the existing predeploys are all hardfork managed) so they would be on a different release schedule

Reply:
> Is this solving specifically for L1 contracts? Any thoughts on L2 contracts?

Moving https://github.com/ethereum-optimism/design-docs/pull/21#issuecomment-2125537843 here too since these are related.

---

Great callout, this is currently unspecified. Having separate `op-contracts-L1/vX.Y.Z` and `op-contracts-L2/vX.Y.Z` releases is a good direction and I'm onboard with that for this proposal.

For cases like custom gas token where both L1 and L2 contracts change, we can reference the corresponding L2 contracts release in the L1 release notes, and vice versa, and note that they must be used together.

I'm pretty sure all of our existing `op-contracts/` releases only released L1 contracts, so we'd also need a plan for the migration. We have a small set of tags, so we could:
- Create `op-contracts-L1/vX.Y.Z` tags at the same commit as all existing `op-contracts/vX.Y.Z`.
- Create releases for them, copy-pasting the existing release notes from the `op-contracts/vX.Y.Z` release into the `op-contracts-L1/vX.Y.Z`
- Update the `op-contracts/vX.Y.Z` to say "This release is deprecated in favor of the identical `op-contracts-L1/vX.Y.Z` release. See here to learn more about contract versioning"

Reply:
Agreed that we will:
- Not change release approach for this design doc. We'll continue doing `op-contracts` level and specify if it's L1 vs. L2. vs. some combination in the release notes
- A future effort will determine best way to handle L2 releases/upgrades in more detail